R/compound_taskid-utils.R

```diff
@@ -0,0 +1,202 @@
+  # compound_taskids set vars will always be set to TRUE in this check.
+  tbl_non_uniq <- tbl[, task_ids] |>
+    purrr::map(~ dplyr::n_distinct(.x) > 1L) |>
+    tibble::as_tibble()
+  tbl_non_uniq[, config_comp_tids] <- TRUE
+
+  # Create a sample (output_type_id) vs compound_taskid membership boolean matrix.
+  # For a task id to be considered a compound_taskid, a unique value must be TRUE
+  # at the sample level and we need to ensure it's not a false positive.
+  # Here we combine the two checks.
+  is_compound_taskid <- apply(spl_uniq[, task_ids], 1,
+    function(x) x & tbl_non_uniq[, task_ids],
+    simplify = FALSE
+  ) |> purrr::reduce(rbind)
+
+  output_type_ids <- spl_uniq[["output_type_id"]]
+
+  tbl_comp_tids <- true_to_names_vector(is_compound_taskid)
+
+  if (length(tbl_comp_tids) > 1L) {
+    error_msg <- "More than 1 unique {.var compound_taskid_set} detected."
+    if (error) {
+      error_bullets <- purrr::map_chr(
+        tbl_comp_tids,
+        .f = function(.x) {
+          paste0("{.val ", .x, "}") |>
+            paste(collapse = ", ") |>
+            stats::setNames("*") |>
+            rlang::format_error_bullets()
+        }
+      )
+      cli::cli_abort(c("x" = error_msg, error_bullets), call = call)
+    }
+    # Attach error attributes
+    attributes(tbl_comp_tids) <- list(
+      errors = get_comp_tid_sample_ids(is_compound_taskid, output_type_ids),
+      msg = cli::format_inline(error_msg, " See `errors` attribute for details.")
+    )
+    return(tbl_comp_tids)
+  }
+
+  tbl_comp_tids <- unlist(tbl_comp_tids)
+
+  if (!all(tbl_comp_tids %in% config_comp_tids)) {
+    invalid_tbl_comp_tids <- setdiff(tbl_comp_tids, config_comp_tids)
+
+    error_vec <- c(
+      "x" = "Finer {.var compound_taskid_set} than allowed detected.",
+      "!" = "{.val {invalid_tbl_comp_tids}} identified as compound task ID{?s} in file but not allowed in config.",
+      "i" = "Compound task IDs should be one of {.val {config_comp_tids}}."
+    )
+
+    if (error) {
+      cli::cli_abort(error_vec, call = call)
+    }
+
+    # Attach error attributes
+    errors <- get_comp_tid_sample_ids(is_compound_taskid, output_type_ids) |>
+      purrr::map(function(.x) {
+        if (any(.x$tbl_comp_tids %in% invalid_tbl_comp_tids)) {
+          c(list(
+            config_comp_tids = config_comp_tids,
+            invalid_tbl_comp_tids = invalid_tbl_comp_tids
+          ), .x)
+        }
+      }) |>
+      purrr::compact()
+
+    attributes(tbl_comp_tids) <- list(
+      errors = errors,
+      msg = cli::format_inline(paste(error_vec, collapse = " "))
+    )
+  }
+  tbl_comp_tids
+}
+
+# Helper functions ----
+# Extract the names of the columns that are TRUE in each row
+true_to_names_vector <- function(x, cols = NULL, unique = TRUE) {
+  if (!is.null(cols)) {
+    x <- x[, cols]
+  }
+  if (unique) {
+    x <- unique(x)
+  }
+  v <- apply(x, 1,
+    function(x) names(x)[x],
+    simplify = FALSE
+  )
```

Comment:
I think this will generate an error if `length(cols) == 1`

Reply:
Thanks! Well caught! I'm so used to working with `tibbles` I've forgotten to check for this behavour! Fixed in 85e42576a392725f9dd3063e3ada4fa713703881